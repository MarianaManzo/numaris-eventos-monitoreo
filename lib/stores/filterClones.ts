import type { EventsFilters, UnitsFilters } from './filterStore';

export const cloneEventsFilters = (filters: EventsFilters): EventsFilters => ({
  ...filters,
  severidades: [...filters.severidades],
  etiquetas: [...filters.etiquetas],
  unidades: [...filters.unidades],
  assignedUsers: [...filters.assignedUsers],
  location: [...filters.location],
  eventTypes: [...filters.eventTypes],
  dateRange: filters.dateRange ? { ...filters.dateRange } : null
});

export const cloneUnitsFilters = (filters: UnitsFilters): UnitsFilters => ({
  ...filters,
  tags: [...filters.tags],
  unidades: [...filters.unidades],
  zones: [...filters.zones],
  zoneTags: [...filters.zoneTags],
  brandModels: [...filters.brandModels],
  status: [...filters.status],
  responsables: [...filters.responsables],
  searchText: filters.searchText,
  lastSeenRange: filters.lastSeenRange ? { ...filters.lastSeenRange } : null
});
