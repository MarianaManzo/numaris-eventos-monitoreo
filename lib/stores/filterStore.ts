import { create } from 'zustand';
import type { EventSeverity } from '@/lib/events/types';

export type FilterDomain = 'events' | 'units';

export type FilterDomainIcon = 'events' | 'units';

export interface AppliedFilter {
  id: string;
  domain: FilterDomain;
  key: string;
  label: string;
  value: string;
  removable: boolean;
  icon?: FilterDomainIcon;
  count?: number;
}

interface DateRange {
  start: string;
  end: string;
}

type EventMultiValueMap = {
  severidades: EventSeverity;
  etiquetas: string;
  unidades: string;
  assignedUsers: string;
  location: string;
  eventTypes: string;
};

type EventMultiValueKey = keyof EventMultiValueMap;

type UnitMultiValueMap = {
  tags: string;
  zones: string;
  brandModels: string;
  status: string;
  responsables: string;
};

type UnitMultiValueKey = keyof UnitMultiValueMap;

export interface EventsFilters {
  estado: 'todos' | 'abiertos' | 'cerrados';
  severidades: EventSeverity[];
  etiquetas: string[];
  unidades: string[];
  dateRange: DateRange | null;
  assignedUsers: string[];
  location: string[];
  eventTypes: string[];
  filterByMapVehicles: boolean;
  focusMode: boolean;
  searchText: string;
}

export interface UnitsFilters {
  tags: string[];
  zones: string[];
  brandModels: string[];
  status: string[];
  lastSeenRange: DateRange | null;
  responsables: string[];
  searchText: string;
}

interface FilterStoreState {
  events: EventsFilters;
  units: UnitsFilters;
  appliedFilters: AppliedFilter[];
  hydrationReady: boolean;
  setEventsFilters: (partial: Partial<EventsFilters>) => void;
  toggleEventFilterValue: <K extends EventMultiValueKey>(key: K, value: EventMultiValueMap[K]) => void;
  setUnitsFilters: (partial: Partial<UnitsFilters>) => void;
  toggleUnitFilterValue: <K extends UnitMultiValueKey>(key: K, value: UnitMultiValueMap[K]) => void;
  removeFilter: (filterId: string) => void;
  clearAllFilters: () => void;
  clearDomainFilters: (domain: FilterDomain) => void;
  hydrateFromQuery: (params?: URLSearchParams | string | null) => void;
  toQueryParams: () => URLSearchParams;
}

const DEFAULT_EVENT_SEVERITIES: EventSeverity[] = ['Alta', 'Media', 'Baja', 'Informativa'];
const DEFAULT_UNIT_STATUS = ['Activo', 'Inactivo', 'En ruta', 'Detenido'];

export const FILTER_QUERY_KEYS = [
  'events_estado',
  'events_severidades',
  'events_etiquetas',
  'events_unidades',
  'events_mapScope',
  'events_focusMode',
  'events_q',
  'units_tags',
  'units_status',
  'units_responsables',
  'units_zones',
  'units_brandModels',
  'units_q'
] as const;

type FilterQueryKey = (typeof FILTER_QUERY_KEYS)[number];

const EVENT_QUERY_KEY_MAP: Record<string, FilterQueryKey> = {
  estado: 'events_estado',
  severidades: 'events_severidades',
  etiquetas: 'events_etiquetas',
  unidades: 'events_unidades',
  filterByMapVehicles: 'events_mapScope',
  focusMode: 'events_focusMode',
  searchText: 'events_q'
};

const UNIT_QUERY_KEY_MAP: Record<string, FilterQueryKey> = {
  tags: 'units_tags',
  status: 'units_status',
  responsables: 'units_responsables',
  zones: 'units_zones',
  brandModels: 'units_brandModels',
  searchText: 'units_q'
};

const createDefaultEventsFilters = (): EventsFilters => ({
  estado: 'todos',
  severidades: [...DEFAULT_EVENT_SEVERITIES],
  etiquetas: [],
  unidades: [],
  dateRange: null,
  assignedUsers: [],
  location: [],
  eventTypes: [],
  filterByMapVehicles: false,
  focusMode: false,
  searchText: ''
});

const createDefaultUnitsFilters = (): UnitsFilters => ({
  tags: [],
  zones: [],
  brandModels: [],
  status: [...DEFAULT_UNIT_STATUS],
  lastSeenRange: null,
  responsables: [],
  searchText: ''
});

const toggleArrayValue = <T>(collection: readonly T[], value: T): T[] =>
  collection.includes(value)
    ? collection.filter((item) => item !== value)
    : [...collection, value];

const removeEventMultiValue = <K extends EventMultiValueKey>(
  events: EventsFilters,
  key: K,
  value: EventMultiValueMap[K]
): EventMultiValueMap[K][] => {
  const filtered = (events[key] as EventMultiValueMap[K][])
    .filter((item) => item !== value);
  events[key] = filtered as EventsFilters[K];
  return filtered;
};

const removeUnitMultiValue = <K extends UnitMultiValueKey>(
  units: UnitsFilters,
  key: K,
  value: UnitMultiValueMap[K]
): UnitMultiValueMap[K][] => {
  const filtered = (units[key] as UnitMultiValueMap[K][])
    .filter((item) => item !== value);
  units[key] = filtered as UnitsFilters[K];
  return filtered;
};

const buildFilterId = (domain: FilterDomain, key: string, value?: string) => {
  const encodedValue = value !== undefined ? encodeURIComponent(value) : '';
  return `${domain}|${key}|${encodedValue}`;
};

const parseFilterId = (filterId: string) => {
  const [domain, key, encodedValue] = filterId.split('|') as [FilterDomain?, string?, string?];
  return {
    domain,
    key,
    value: encodedValue ? decodeURIComponent(encodedValue) : undefined
  };
};

const arraysHaveSameMembers = (a: string[], b: string[]) => {
  if (a.length !== b.length) return false;
  const setA = new Set(a);
  return b.every((value) => setA.has(value));
};

const buildAppliedFilters = (events: EventsFilters, units: UnitsFilters): AppliedFilter[] => {
  const pills: AppliedFilter[] = [];

  // Events filters
  if (events.estado !== 'todos') {
    pills.push({
      id: buildFilterId('events', 'estado', events.estado),
      domain: 'events',
      key: 'estado',
      label: 'Estado',
      value: events.estado === 'abiertos' ? 'Abiertos' : 'Cerrados',
      removable: true,
      icon: 'events'
    });
  }

  if (!arraysHaveSameMembers(events.severidades, DEFAULT_EVENT_SEVERITIES)) {
    events.severidades.forEach((severity) => {
      pills.push({
        id: buildFilterId('events', 'severidades', severity),
        domain: 'events',
        key: 'severidades',
        label: 'Severidad',
        value: severity,
        removable: true,
        icon: 'events'
      });
    });
  }

  events.etiquetas.forEach((etiqueta) => {
    pills.push({
      id: buildFilterId('events', 'etiquetas', etiqueta),
      domain: 'events',
      key: 'etiquetas',
      label: 'Etiqueta',
      value: etiqueta,
      removable: true,
      icon: 'events'
    });
  });

  events.unidades.forEach((unidad) => {
    pills.push({
      id: buildFilterId('events', 'unidades', unidad),
      domain: 'events',
      key: 'unidades',
      label: 'Unidad',
      value: unidad,
      removable: true,
      icon: 'events'
    });
  });

  if (events.filterByMapVehicles) {
    pills.push({
      id: buildFilterId('events', 'filterByMapVehicles', 'visible'),
      domain: 'events',
      key: 'filterByMapVehicles',
      label: 'Mapa',
      value: 'Solo visibles',
      removable: true,
      icon: 'events'
    });
  }

  if (events.focusMode) {
    pills.push({
      id: buildFilterId('events', 'focusMode', 'on'),
      domain: 'events',
      key: 'focusMode',
      label: 'Focus mode',
      value: 'Activo',
      removable: true,
      icon: 'events'
    });
  }

  if (events.searchText.trim().length > 0) {
    pills.push({
      id: buildFilterId('events', 'searchText', events.searchText.trim()),
      domain: 'events',
      key: 'searchText',
      label: 'Búsqueda',
      value: events.searchText.trim(),
      removable: true,
      icon: 'events'
    });
  }

  // Units filters
  units.tags.forEach((tag) => {
    pills.push({
      id: buildFilterId('units', 'tags', tag),
      domain: 'units',
      key: 'tags',
      label: 'Etiqueta',
      value: tag,
      removable: true,
      icon: 'units'
    });
  });

  if (!arraysHaveSameMembers(units.status, DEFAULT_UNIT_STATUS)) {
    units.status.forEach((status) => {
      pills.push({
        id: buildFilterId('units', 'status', status),
        domain: 'units',
        key: 'status',
        label: 'Estado',
        value: status,
        removable: true,
        icon: 'units'
      });
    });
  }

  units.responsables.forEach((responsable) => {
    pills.push({
      id: buildFilterId('units', 'responsables', responsable),
      domain: 'units',
      key: 'responsables',
      label: 'Responsable',
      value: responsable,
      removable: true,
      icon: 'units'
    });
  });

  units.zones.forEach((zone) => {
    pills.push({
      id: buildFilterId('units', 'zones', zone),
      domain: 'units',
      key: 'zones',
      label: 'Zona',
      value: zone,
      removable: true,
      icon: 'units'
    });
  });

  units.brandModels.forEach((brandModel) => {
    pills.push({
      id: buildFilterId('units', 'brandModels', brandModel),
      domain: 'units',
      key: 'brandModels',
      label: 'Modelo',
      value: brandModel,
      removable: true,
      icon: 'units'
    });
  });

  if (units.searchText.trim().length > 0) {
    pills.push({
      id: buildFilterId('units', 'searchText', units.searchText.trim()),
      domain: 'units',
      key: 'searchText',
      label: 'Búsqueda',
      value: units.searchText.trim(),
      removable: true,
      icon: 'units'
    });
  }

  // Aggregate duplicates to support +N indicator
  const aggregated = new Map<string, AppliedFilter>();
  pills.forEach((pill) => {
    const existing = aggregated.get(pill.id);
    if (existing) {
      const nextCount = (existing.count ?? 1) + 1;
      aggregated.set(pill.id, {
        ...existing,
        count: nextCount
      });
    } else {
      aggregated.set(pill.id, pill);
    }
  });

  return Array.from(aggregated.values());
};

const buildQueryFromFilters = (events: EventsFilters, units: UnitsFilters) => {
  const params = new URLSearchParams();

  if (events.estado !== 'todos') {
    params.set(EVENT_QUERY_KEY_MAP.estado, events.estado);
  }

  if (!arraysHaveSameMembers(events.severidades, DEFAULT_EVENT_SEVERITIES) && events.severidades.length > 0) {
    params.set(EVENT_QUERY_KEY_MAP.severidades, events.severidades.join(','));
  }

  if (events.etiquetas.length > 0) {
    params.set(EVENT_QUERY_KEY_MAP.etiquetas, events.etiquetas.join(','));
  }

  if (events.unidades.length > 0) {
    params.set(EVENT_QUERY_KEY_MAP.unidades, events.unidades.join(','));
  }

  if (events.filterByMapVehicles) {
    params.set(EVENT_QUERY_KEY_MAP.filterByMapVehicles, 'visible');
  }

  if (events.focusMode) {
    params.set(EVENT_QUERY_KEY_MAP.focusMode, 'on');
  }

  if (events.searchText.trim().length > 0) {
    params.set(EVENT_QUERY_KEY_MAP.searchText, events.searchText.trim());
  }

  if (units.tags.length > 0) {
    params.set(UNIT_QUERY_KEY_MAP.tags, units.tags.join(','));
  }

  if (!arraysHaveSameMembers(units.status, DEFAULT_UNIT_STATUS) && units.status.length > 0) {
    params.set(UNIT_QUERY_KEY_MAP.status, units.status.join(','));
  }

  if (units.responsables.length > 0) {
    params.set(UNIT_QUERY_KEY_MAP.responsables, units.responsables.join(','));
  }

  if (units.zones.length > 0) {
    params.set(UNIT_QUERY_KEY_MAP.zones, units.zones.join(','));
  }

  if (units.brandModels.length > 0) {
    params.set(UNIT_QUERY_KEY_MAP.brandModels, units.brandModels.join(','));
  }

  if (units.searchText.trim().length > 0) {
    params.set(UNIT_QUERY_KEY_MAP.searchText, units.searchText.trim());
  }

  return params;
};

const parseSearchParams = (input?: URLSearchParams | string | null) => {
  if (!input) {
    return new URLSearchParams();
  }
  if (typeof input === 'string') {
    return new URLSearchParams(input);
  }
  return new URLSearchParams(input.toString());
};

export const useFilterStore = create<FilterStoreState>((set, get) => ({
  events: createDefaultEventsFilters(),
  units: createDefaultUnitsFilters(),
  appliedFilters: [],
  hydrationReady: false,

  setEventsFilters: (partial) => {
    set((state) => {
      const events = { ...state.events, ...partial };
      return {
        events,
        appliedFilters: buildAppliedFilters(events, state.units)
      };
    });
  },

  toggleEventFilterValue: <K extends EventMultiValueKey>(key: K, value: EventMultiValueMap[K]) => {
    set((state) => {
      const current = state.events[key] as EventMultiValueMap[K][];
      if (!current) {
        return {};
      }
      const nextValues = toggleArrayValue(current, value);
      const events = {
        ...state.events,
        [key]: nextValues
      } as EventsFilters;
      return {
        events,
        appliedFilters: buildAppliedFilters(events, state.units)
      };
    });
  },

  setUnitsFilters: (partial) => {
    set((state) => {
      const units = { ...state.units, ...partial };
      return {
        units,
        appliedFilters: buildAppliedFilters(state.events, units)
      };
    });
  },

  toggleUnitFilterValue: <K extends UnitMultiValueKey>(key: K, value: UnitMultiValueMap[K]) => {
    set((state) => {
      const current = state.units[key] as UnitMultiValueMap[K][];
      if (!current) {
        return {};
      }
      const nextValues = toggleArrayValue(current, value);
      const units = {
        ...state.units,
        [key]: nextValues
      } as UnitsFilters;
      return {
        units,
        appliedFilters: buildAppliedFilters(state.events, units)
      };
    });
  },

  removeFilter: (filterId) => {
    const { domain, key, value } = parseFilterId(filterId);
    if (!domain || !key) {
      return;
    }

    if (domain === 'events') {
      set((state) => {
        const events = { ...state.events };
        switch (key) {
          case 'estado':
            events.estado = 'todos';
            break;
          case 'severidades':
            if (value) {
              const next = removeEventMultiValue(events, 'severidades', value as EventSeverity);
              events.severidades = next.length > 0 ? next : [...DEFAULT_EVENT_SEVERITIES];
            }
            break;
          case 'responsables':
          case 'etiquetas':
          case 'unidades':
          case 'assignedUsers':
          case 'location':
          case 'eventTypes':
            if (value) {
              removeEventMultiValue(
                events,
                key as EventMultiValueKey,
                value as EventMultiValueMap[EventMultiValueKey]
              );
            }
            break;
          case 'filterByMapVehicles':
            events.filterByMapVehicles = false;
            break;
          case 'focusMode':
            events.focusMode = false;
            break;
          case 'searchText':
            events.searchText = '';
            break;
          default:
            break;
        }
        return {
          events,
          appliedFilters: buildAppliedFilters(events, state.units)
        };
      });
      return;
    }

    if (domain === 'units') {
      set((state) => {
        const units = { ...state.units };
        switch (key) {
          case 'tags':
          case 'zones':
          case 'brandModels':
          case 'responsables':
            if (value) {
              removeUnitMultiValue(
                units,
                key as UnitMultiValueKey,
                value as UnitMultiValueMap[UnitMultiValueKey]
              );
            }
            break;
          case 'status':
            if (value) {
              const next = removeUnitMultiValue(units, 'status', value as UnitMultiValueMap['status']);
              units.status = next.length > 0 ? next : [...DEFAULT_UNIT_STATUS];
            }
            break;
          case 'searchText':
            units.searchText = '';
            break;
          default:
            break;
        }

        return {
          units,
          appliedFilters: buildAppliedFilters(state.events, units)
        };
      });
    }
  },

  clearAllFilters: () => {
    const events = createDefaultEventsFilters();
    const units = createDefaultUnitsFilters();
    set({
      events,
      units,
      appliedFilters: buildAppliedFilters(events, units)
    });
  },

  clearDomainFilters: (domain) => {
    if (domain === 'events') {
      const events = createDefaultEventsFilters();
      set((state) => ({
        events,
        appliedFilters: buildAppliedFilters(events, state.units)
      }));
      return;
    }

    if (domain === 'units') {
      const units = createDefaultUnitsFilters();
      set((state) => ({
        units,
        appliedFilters: buildAppliedFilters(state.events, units)
      }));
    }
  },

  hydrateFromQuery: (input) => {
    const params = parseSearchParams(input);
    const hasFilterParams = FILTER_QUERY_KEYS.some((key) => params.has(key));

    if (!hasFilterParams) {
      // Still mark hydration ready so subsequent sync works
      set((state) => ({
        appliedFilters: buildAppliedFilters(state.events, state.units),
        hydrationReady: true
      }));
      return;
    }

    const events = createDefaultEventsFilters();
    const units = createDefaultUnitsFilters();

    const estadoParam = params.get(EVENT_QUERY_KEY_MAP.estado);
    if (estadoParam === 'abiertos' || estadoParam === 'cerrados') {
      events.estado = estadoParam;
    }

    const severidadesParam = params.get(EVENT_QUERY_KEY_MAP.severidades);
    if (severidadesParam) {
      const severidades = severidadesParam.split(',').map((value) => value.trim()).filter(Boolean) as EventSeverity[];
      if (severidades.length > 0) {
        events.severidades = severidades;
      }
    }

    const responsablesParam = params.get(EVENT_QUERY_KEY_MAP.responsables);
    if (responsablesParam) {
      events.responsables = responsablesParam.split(',').map((value) => value.trim()).filter(Boolean);
    }

    const etiquetasParam = params.get(EVENT_QUERY_KEY_MAP.etiquetas);
    if (etiquetasParam) {
      events.etiquetas = etiquetasParam.split(',').map((value) => value.trim()).filter(Boolean);
    }

    const unidadesParam = params.get(EVENT_QUERY_KEY_MAP.unidades);
    if (unidadesParam) {
      events.unidades = unidadesParam.split(',').map((value) => value.trim()).filter(Boolean);
    }

    const mapScopeParam = params.get(EVENT_QUERY_KEY_MAP.filterByMapVehicles);
    if (mapScopeParam === 'visible') {
      events.filterByMapVehicles = true;
    }

    const focusModeParam = params.get(EVENT_QUERY_KEY_MAP.focusMode);
    if (focusModeParam === 'on') {
      events.focusMode = true;
    }

    const eventsSearch = params.get(EVENT_QUERY_KEY_MAP.searchText);
    if (eventsSearch) {
      events.searchText = eventsSearch;
    }

    const tagsParam = params.get(UNIT_QUERY_KEY_MAP.tags);
    if (tagsParam) {
      units.tags = tagsParam.split(',').map((value) => value.trim()).filter(Boolean);
    }

    const statusParam = params.get(UNIT_QUERY_KEY_MAP.status);
    if (statusParam) {
      const statusValues = statusParam.split(',').map((value) => value.trim()).filter(Boolean);
      if (statusValues.length > 0) {
        units.status = statusValues;
      }
    }

    const unitResponsablesParam = params.get(UNIT_QUERY_KEY_MAP.responsables);
    if (unitResponsablesParam) {
      units.responsables = unitResponsablesParam.split(',').map((value) => value.trim()).filter(Boolean);
    }

    const zonesParam = params.get(UNIT_QUERY_KEY_MAP.zones);
    if (zonesParam) {
      units.zones = zonesParam.split(',').map((value) => value.trim()).filter(Boolean);
    }

    const brandModelsParam = params.get(UNIT_QUERY_KEY_MAP.brandModels);
    if (brandModelsParam) {
      units.brandModels = brandModelsParam.split(',').map((value) => value.trim()).filter(Boolean);
    }

    const unitsSearch = params.get(UNIT_QUERY_KEY_MAP.searchText);
    if (unitsSearch) {
      units.searchText = unitsSearch;
    }

    set({
      events,
      units,
      appliedFilters: buildAppliedFilters(events, units),
      hydrationReady: true
    });
  },

  toQueryParams: () => {
    const { events, units } = get();
    return buildQueryFromFilters(events, units);
  }
}));
