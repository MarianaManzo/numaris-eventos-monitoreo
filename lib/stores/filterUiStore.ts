'use client';

import { create } from 'zustand';

type DropdownKey = 'units' | 'events' | 'zones' | null;

type DropdownSetter = DropdownKey | ((current: DropdownKey) => DropdownKey);

interface FilterUiState {
  isBarOpen: boolean;
  activeDropdown: DropdownKey;
  openBar: () => void;
  closeBar: () => void;
  toggleBar: () => void;
  setActiveDropdown: (key: DropdownSetter) => void;
}

export const useFilterUiStore = create<FilterUiState>((set) => ({
  isBarOpen: false,
  activeDropdown: null,
  openBar: () => set({ isBarOpen: true }),
  closeBar: () => set({ isBarOpen: false, activeDropdown: null }),
  toggleBar: () =>
    set((state) => ({
      isBarOpen: !state.isBarOpen,
      activeDropdown: state.isBarOpen ? null : state.activeDropdown
    })),
  setActiveDropdown: (key) =>
    set((state) => ({
      activeDropdown: typeof key === 'function' ? key(state.activeDropdown) : key
    }))
}));
