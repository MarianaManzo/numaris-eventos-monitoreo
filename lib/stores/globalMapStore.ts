import { create } from 'zustand';

/**
 * Global Map Store - Cross-View Context Sharing
 *
 * Manages layer visibility and map state that persists across
 * Vehicles, Events, and Zonas views.
 *
 * Design Philosophy:
 * - Layer visibility is a user's explicit choice about map clutter
 * - Should persist when switching between monitoring contexts
 * - Enables spatial awareness and geographic analysis
 */

interface GlobalMapStore {
  // ===== Phase 1: Layer Visibility =====

  /**
   * Show/hide vehicles layer across all views
   * - In Unidades view: Primary layer (100% opacity)
   * - In Eventos/Zonas views: Context layer (70% opacity)
   */
  showVehiclesOnMap: boolean;

  /**
   * Show/hide events layer across all views
   * - In Eventos view: Primary layer (100% opacity)
   * - In Unidades/Zonas views: Context layer (70% opacity)
   */
  showEventsOnMap: boolean;

  /**
   * Show/hide zonas layer across all views
   * - In Zonas view: Primary layer (100% opacity)
   * - In Unidades/Eventos views: Context layer (70% opacity)
   */
  showZonasOnMap: boolean;

  /**
   * Show/hide vehicle labels across all views
   */
  showVehicleLabels: boolean;

  /**
   * Show/hide event labels across all views
   */
  showEventLabels: boolean;

  /**
   * Show/hide zona labels across all views
   */
  showZonaLabels: boolean;

  // ===== Actions =====

  /**
   * Toggle vehicles layer visibility
   * @param show - true to show vehicles, false to hide
   */
  setShowVehiclesOnMap: (show: boolean) => void;

  /**
   * Toggle events layer visibility
   * @param show - true to show events, false to hide
   */
  setShowEventsOnMap: (show: boolean) => void;

  /**
   * Toggle zonas layer visibility
   * @param show - true to show zonas, false to hide
   */
  setShowZonasOnMap: (show: boolean) => void;

  /**
   * Toggle vehicle labels visibility
   */
  setShowVehicleLabels: (show: boolean) => void;

  /**
   * Toggle event labels visibility
   */
  setShowEventLabels: (show: boolean) => void;

  /**
   * Toggle zona labels visibility
   */
  setShowZonaLabels: (show: boolean) => void;

  /**
   * Convenience method to show all layers
   */
  showAllLayers: () => void;

  /**
   * Convenience method to hide all layers
   */
  hideAllLayers: () => void;
}

/**
 * Global Map Store Hook
 *
 * Usage:
 * ```typescript
 * const { showVehiclesOnMap, setShowVehiclesOnMap } = useGlobalMapStore();
 *
 * // In any map view
 * {showVehiclesOnMap && (
 *   <ClusteredVehicleMarkers
 *     opacity={isPrimaryView ? 1.0 : 0.7}
 *     size={isPrimaryView ? 'normal' : 'small'}
 *   />
 * )}
 * ```
 */
export const useGlobalMapStore = create<GlobalMapStore>((set) => ({
  // Default: all layers visible
  // Rationale: Users should see full context by default,
  // can selectively hide layers to reduce clutter
  showVehiclesOnMap: true,
  showEventsOnMap: true,
  showZonasOnMap: true,
  showVehicleLabels: true,
  showEventLabels: true,
  showZonaLabels: true,

  // Actions
  setShowVehiclesOnMap: (show) => set({ showVehiclesOnMap: show }),
  setShowEventsOnMap: (show) => set({ showEventsOnMap: show }),
  setShowZonasOnMap: (show) => set({ showZonasOnMap: show }),
  setShowVehicleLabels: (show) => set({ showVehicleLabels: show }),
  setShowEventLabels: (show) => set({ showEventLabels: show }),
  setShowZonaLabels: (show) => set({ showZonaLabels: show }),

  // Convenience methods
  showAllLayers: () => set({
    showVehiclesOnMap: true,
    showEventsOnMap: true,
    showZonasOnMap: true,
    showVehicleLabels: true,
    showEventLabels: true,
    showZonaLabels: true
  }),

  hideAllLayers: () => set({
    showVehiclesOnMap: false,
    showEventsOnMap: false,
    showZonasOnMap: false,
    showVehicleLabels: false,
    showEventLabels: false,
    showZonaLabels: false
  })
}));
