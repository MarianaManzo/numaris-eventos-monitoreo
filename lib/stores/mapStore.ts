import { create } from 'zustand';

interface MapStore {
  mapCenter: [number, number] | null;
  mapZoom: number | null;
  lastUpdateTime: number;
  setMapView: (center: [number, number], zoom: number) => void;
  clearMapView: () => void;
  shouldRestoreView: () => boolean;
}

export const useMapStore = create<MapStore>((set, get) => ({
  mapCenter: null,
  mapZoom: null,
  lastUpdateTime: 0,

  setMapView: (center, zoom) => {
    set({
      mapCenter: center,
      mapZoom: zoom,
      lastUpdateTime: Date.now()
    });
  },

  clearMapView: () => {
    set({
      mapCenter: null,
      mapZoom: null,
      lastUpdateTime: 0
    });
  },

  shouldRestoreView: () => {
    const state = get();
    // Restore if view was saved in last 30 seconds (increased for tab switches)
    const timeSinceUpdate = Date.now() - state.lastUpdateTime;
    return state.mapCenter !== null && state.mapZoom !== null && timeSinceUpdate < 30000;
  }
}));