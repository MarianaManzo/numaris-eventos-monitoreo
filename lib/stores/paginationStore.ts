import { create } from 'zustand';

export type PaginationDomain = 'units' | 'events' | 'zones';

interface PaginationState {
  page: Record<PaginationDomain, number>;
  pageSize: Record<PaginationDomain, number>;
  setPage: (domain: PaginationDomain, page: number) => void;
  setPageSize: (domain: PaginationDomain, size: number) => void;
  resetPage: (domain: PaginationDomain) => void;
  resetAll: () => void;
}

const DEFAULT_PAGE_SIZE = 10;

const initialPageState: PaginationState['page'] = {
  units: 0,
  events: 0,
  zones: 0
};

const initialPageSizeState: PaginationState['pageSize'] = {
  units: DEFAULT_PAGE_SIZE,
  events: DEFAULT_PAGE_SIZE,
  zones: DEFAULT_PAGE_SIZE
};

export const usePaginationStore = create<PaginationState>((set) => ({
  page: initialPageState,
  pageSize: initialPageSizeState,
  setPage: (domain, page) =>
    set((state) => ({
      page: {
        ...state.page,
        [domain]: Math.max(0, page)
      }
    })),
  setPageSize: (domain, size) =>
    set((state) => ({
      pageSize: {
        ...state.pageSize,
        [domain]: Math.max(1, size)
      },
      page: {
        ...state.page,
        [domain]: 0
      }
    })),
  resetPage: (domain) =>
    set((state) => ({
      page: {
        ...state.page,
        [domain]: 0
      }
    })),
  resetAll: () =>
    set({
      page: initialPageState,
      pageSize: initialPageSizeState
    })
}));
