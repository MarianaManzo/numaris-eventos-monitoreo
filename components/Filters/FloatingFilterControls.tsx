'use client';

import { useMemo, MouseEvent } from 'react';
import { Button, Dropdown, Space, Tag } from 'antd';
import { FunnelSimple, Truck } from 'phosphor-react';
import { useFilterStore } from '@/lib/stores/filterStore';

const DOMAIN_CONFIG = [
  {
    key: 'units' as const,
    label: 'Unidades',
    icon: <Truck size={16} />
  },
  {
    key: 'events' as const,
    label: 'Eventos',
    icon: <FunnelSimple size={16} />
  }
] satisfies Array<{
  key: 'events' | 'units';
  label: string;
  icon: React.ReactNode;
}>;

export default function FloatingFilterControls() {
  const appliedFilters = useFilterStore((state) => state.appliedFilters);
  const removeFilter = useFilterStore((state) => state.removeFilter);
  const clearAllFilters = useFilterStore((state) => state.clearAllFilters);

  const grouped = useMemo(() => {
    return appliedFilters.reduce<Record<'events' | 'units', typeof appliedFilters>>(
      (acc, filter) => {
        if (filter.domain === 'events' || filter.domain === 'units') {
          acc[filter.domain] = [...acc[filter.domain], filter];
        }
        return acc;
      },
      {
        events: [],
        units: []
      }
    );
  }, [appliedFilters]);

  const totalFilters = grouped.events.length + grouped.units.length;
  if (totalFilters === 0) {
    return null;
  }

  const handleRemove =
    (filterId: string) =>
    (event: MouseEvent<HTMLButtonElement>) => {
      event.preventDefault();
      event.stopPropagation();
      removeFilter(filterId);
    };

  return (
    <div className="floating-filter-controls">
      <Space className="floating-filter-button-group">
        {DOMAIN_CONFIG.map(({ key, label, icon }) => {
          const filters = grouped[key];
          if (filters.length === 0) {
            return null;
          }

          const items = filters.map((filter) => ({
            key: filter.id,
            label: (
              <div className="floating-filter-item">
                <span className="floating-filter-item__text">
                  {renderFilterValue(key, filter)}
                  {filter.count && filter.count > 1 && (
                    <span className="floating-filter-item__count">(+{filter.count - 1})</span>
                  )}
                </span>
                <Button
                  type="text"
                  size="small"
                  className="floating-filter-remove"
                  onClick={handleRemove(filter.id)}
                >
                  Ã—
                </Button>
              </div>
            )
          }));

          return (
            <Dropdown key={key} menu={{ items }} trigger={['click']} placement="bottomLeft">
              <Button className="floating-filter-button" icon={icon}>
                {label}
                <Tag className="floating-filter-button__tag">{filters.length}</Tag>
                <svg
                  className="floating-filter-button__caret"
                  width={18}
                  height={18}
                  viewBox="0 0 24 24"
                  fill="currentColor"
                  aria-hidden="true"
                >
                  <path d="M7 10c-.552 0-.834.633-.44 1.026l4.293 4.293a1 1 0 0 0 1.414 0L16.56 11.026C16.953 10.633 16.671 10 16.118 10H7z" />
                </svg>
              </Button>
            </Dropdown>
          );
        })}
      </Space>
      <Button
        type="default"
        size="middle"
        className="floating-filter-clear"
        onClick={clearAllFilters}
      >
        Limpiar todo
      </Button>
    </div>
  );
}

const renderFilterValue = (
  domain: 'events' | 'units',
  filter: { key: string; value: string; label: string }
) => {
  if (domain === 'units' && filter.key === 'unidades') {
    return (
      <div className="floating-filter-unit-layout">
        <span className="floating-filter-unit-label">{filter.label}:</span>
        <span className="floating-filter-unit-value">
          <span className="floating-filter-unit-icon" aria-hidden="true">
            <svg viewBox="0 0 24 24" fill="none">
              <rect x="1" y="9" width="22" height="8" rx="2" fill="#4ade80" />
              <path d="M17 8H7l1.2-2.4A2 2 0 0 1 10 4h4a2 2 0 0 1 1.8 1.1L17 8Z" fill="#4ade80" />
              <circle cx="7" cy="17" r="2" fill="#ffffff" />
              <circle cx="17" cy="17" r="2" fill="#ffffff" />
            </svg>
          </span>
          {filter.value}
        </span>
      </div>
    );
  }

  if (domain === 'events' && filter.key === 'estado') {
    return (
      <>
        {filter.label}:{' '}
        <span className={
          filter.value === 'abiertos'
            ? 'floating-filter-state floating-filter-state--open'
            : 'floating-filter-state floating-filter-state--closed'
        }>
          <span className="floating-filter-state__dot" />
          {filter.value === 'abiertos' ? 'Abierto' : 'Cerrado'}
        </span>
      </>
    );
  }

  if (domain === 'events' && filter.key === 'severidades') {
    const severityClass = `floating-filter-severity floating-filter-severity--${filter.value.toLowerCase()}`;
    return (
      <>
        {filter.label}:{' '}
        <span className={severityClass}>{filter.value}</span>
      </>
    );
  }

  return (
    <>
      {filter.label}: {filter.value}
    </>
  );
};
