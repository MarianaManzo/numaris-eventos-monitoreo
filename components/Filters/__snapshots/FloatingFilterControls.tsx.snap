/* placeholder snapshot to keep directory creation if needed */

